auth/gcp/authenticator.go

```diff
@@ -0,0 +1,422 @@
+		KMSKeyName string
+		// UnsafeState can be used to skip the encryption of the "state" token
+		// within the auth flow.
+		UnsafeState bool
+
+		// AuthConfig is used by Authenticator.Middleware and callback to enable the
+		// Google OAuth flow.
+		AuthConfig *oauth2.Config
+
+		// HeaderExceptions can optionally be included. Any requests that include any of
+		// the headers included will skip all Authenticator.Middlware checks and no
+		// claims information will be added to the context.
+		// This can be useful for unspoofable headers like Google App Engine's
+		// "X-AppEngine-*" headers for Google Task Queues.
+		HeaderExceptions []string
+
+		// CustomExceptionsFunc allows any custom exceptions based on the request. For
+		// example, looking for specific URIs.  Return true if should be allowed.  If
+		// false is returned, normal cookie-based authentication happens.
+		CustomExceptionsFunc func(context.Context, *http.Request) bool
+
+		// IDConfig will be used to verify the Google Identity JWS when it is inbound
+		// in the HTTP cookie.
+		IDConfig IdentityConfig
+		// IDVerifyFunc allows developers to add their own verification on the user
+		// claims. For example, one could enable access for anyone with an email domain
+		// of "@example.com".
+		IDVerifyFunc func(context.Context, IdentityClaimSet) bool
+
+		// Logger will be used to log any errors encountered during the auth flow.
+		Logger log.Logger
+	}
+)
+
+// NewAuthenticator will instantiate a new Authenticator, which can be used for verifying
+// a number of authentication styles within the Google Cloud Platform ecosystem.
+func NewAuthenticator(ctx context.Context, cfg AuthenticatorConfig) (Authenticator, error) {
+	ks, err := NewIdentityPublicKeySource(ctx, cfg.IDConfig)
+	if err != nil {
+		return Authenticator{}, errors.Wrap(err, "unable to init key source")
+	}
+	u, err := url.Parse(cfg.AuthConfig.RedirectURL)
+	if err != nil {
+		return Authenticator{}, errors.Wrap(err, "unable to pasrse redirect URL")
+	}
+	var keys *kms.KeyManagementClient
+	if !cfg.UnsafeState {
+		keys, err = kms.NewKeyManagementClient(ctx)
+		if err != nil {
+			return Authenticator{}, errors.Wrap(err, "unable to init KMS client")
+		}
+	}
+	if cfg.Logger == nil {
+		cfg.Logger = log.NewNopLogger()
+	}
+	return Authenticator{
+		cfg:          cfg,
+		keys:         keys,
+		cookieDomain: strings.Split(u.Host, ":")[0],
+		secureCookie: u.Scheme == "https",
+		callbackPath: u.Path,
+		verifier: auth.NewVerifier(ks, IdentityClaimsDecoderFunc,
+			IdentityVerifyFunc(cfg.IDVerifyFunc)),
+	}, nil
+}
+
+// LogOut can be used to clear an existing session. It will add an HTTP cookie with a -1
+// "MaxAge" to the response to remove the cookie from the logged in user's browser.
+func (c Authenticator) LogOut(w http.ResponseWriter) {
+	http.SetCookie(w, &http.Cookie{
+		Name:    c.cfg.CookieName,
+		Domain:  c.cookieDomain,
+		Value:   "",
+		MaxAge:  -1,
+		Expires: time.Unix(0, 0),
```

Comment:
You need to set Path to `/` otherwise cookie doesn't get cleared by browser.

Reply:
Also think it makes sense to also set `secure`  flag to match the originating set cookie.